flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java

```diff
@@ -0,0 +1,340 @@
+ *
+ * <p>KBinsDiscretizer implements three different binning strategies, and it can be set by {@link
+ * KBinsDiscretizerParams#STRATEGY}. If the strategy is set as {@link KBinsDiscretizerParams#KMEANS}
+ * or {@link KBinsDiscretizerParams#QUANTILE}, users should further set {@link
+ * KBinsDiscretizerParams#SUB_SAMPLES} for better performance.
+ *
+ * <p>There are several corner cases for different inputs as listed below:
+ *
+ * <ul>
+ *   <li>When the input values of one column are all the same, then they should be mapped to the
+ *       same bin (i.e., the zero-th bin). Thus the corresponding bin edges are `{Double.MIN_VALUE,
+ *       Double.MAX_VALUE}`.
+ *   <li>When the number of distinct values of one column is less than the specified number of bins
+ *       and the {@link KBinsDiscretizerParams#STRATEGY} is set as {@link
+ *       KBinsDiscretizerParams#KMEANS}, we switch to {@link KBinsDiscretizerParams#UNIFORM}.
+ *   <li>When the width of one output bin is zero, i.e., the left edge equals to the right edge of
+ *       the bin, we remove it.
+ * </ul>
+ */
+public class KBinsDiscretizer
+        implements Estimator<KBinsDiscretizer, KBinsDiscretizerModel>,
+                KBinsDiscretizerParams<KBinsDiscretizer> {
+    private static final Logger LOG = LoggerFactory.getLogger(KBinsDiscretizer.class);
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public KBinsDiscretizer() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public KBinsDiscretizerModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        String inputCol = getInputCol();
+        String strategy = getStrategy();
+        int numBins = getNumBins();
+
+        DataStream<DenseVector> inputData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                (MapFunction<Row, DenseVector>)
+                                        value -> ((Vector) value.getField(inputCol)).toDense());
+
+        DataStream<DenseVector> preprocessedData;
+        if (strategy.equals(UNIFORM)) {
+            preprocessedData =
+                    inputData
+                            .transform(
+                                    "reduceInEachPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .transform(
+                                    "reduceInFinalPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .setParallelism(1);
+        } else {
+            preprocessedData = DataStreamUtils.sample(inputData, getSubSamples(), 2022L);
```

Comment:
Would it be better to change the random seed from `2022L` to `getClass().getName().hashCode()`? If one day we would support the random seed parameter, this would be its default value.